src/lib.rs

```diff
@@ -476,16 +478,20 @@ impl<'de> Deserializer<'de> {
         }
 
         unsafe {
-            input_buffer
-                .as_mut_slice()
-                .get_unchecked_mut(..len)
-                .clone_from_slice(input);
-            *(input_buffer.get_unchecked_mut(len)) = 0;
-            input_buffer.set_len(len);
+            std::ptr::copy_nonoverlapping(
+                input.as_ptr(),
+                input_buffer.as_mut_ptr(),
+                len,
+            );
+
+            let to_fill = input_buffer.capacity() - len;
+            std::ptr::write_bytes(input_buffer.as_mut_ptr().add(len), 0, to_fill);
```

Comment:
No, because if you just write one byte, you get a use-of-uninitialized-value from memory sanitizer, and I'm sure we're reading uninit bytes into a `u64` (say, read_true_atom run on `[t r u e \x00 <uninit> <uninit> <uninit>]`

Simply reading uninit bytes into a `u64` is UB (see: https://github.com/rust-lang/unsafe-code-guidelines/issues/71), and memory sanitizer complaining means there's *definitely* UB here (memory sanitizer is a very conservative check, only checking for *branching* on uninit bytes).

Upstream seems to... just ignore the problem (they run memory sanitizer, but ignore this).

We presumably only need to ensure that `input.len() + SIMDJSON_PADDING` bytes are zero filled, so the user could re-use buffers manually (which would remove both the allocation and the zeroing). As in, the write_bytes could only be done if we're using a buffer that isn't long enough (in `len`, not `capacity`).

Reply:
They ignore it because it's not a real problem, it's a "UB by decree" not because any behavior is actually undefined, the discussion you linked says this quite clearly.

There are no possible invalid combinations of bits in a  u64. Rust has the tendency to be overly careful with safety, which is great, but sometimes it's worth understanding what happens under the hood - that's why unsafe exists, after all, the compiler can't always make the right call as it misses context, like that' we're not reading a u64 but a SIMD register.

But it's not worth arguing over this, criterion says the change has no significant impact on performance, some tests are marginally faster others marginally slower - it's with run on run varriance. So there is no reason to keep potentially UB behaviour just for the sake of it, that'd be silly, if we can be compiler-agreed-safe and fast that's definitely the route to take :D

Reply:
> But it's not worth arguing over this

I'm not commenting here to argue with you, but to note that you should be significantly more careful with uninitialized integers since https://github.com/rust-lang/rust/pull/106294